lib/Map/Vector/ProtomapsGeojsonSource.ts

```diff
@@ -82,73 +82,133 @@ export class ProtomapsGeojsonSource implements TileSource {
+        geojsonVtTileToProtomapsFeatures(tile.features, tileSize)
       );
     }
 
     return result;
   }
 }
+
+export const geomTypeMap = (
+  type: string | null | undefined
+): GeomType | null => {
+  switch (type) {
+    case "Point":
+    case "MultiPoint":
+      return GeomType.Point;
+    case "LineString":
+    case "MultiLineString":
+      return GeomType.Line;
+    case "Polygon":
+    case "MultiPolygon":
+      return GeomType.Polygon;
+    default:
+      return null;
+  }
+};
+
+export function geojsonVtTileToProtomapsFeatures(
+  features: geojsonvt.Feature[],
+  tileSize: number
+): ProtomapsFeature[] {
+  const scale = tileSize / GEOJSON_VT_EXTENT;
+
+  return features
+    .map((f) => {
+      let transformedGeom: Point[][] = [];
+      let numVertices = 0;
+
+      // Calculate bbox
+      const bbox: Bbox = {
+        minX: Infinity,
+        minY: Infinity,
+        maxX: -Infinity,
+        maxY: -Infinity
+      };
+
+      // Multi-polygon
+      if (Array.isArray(f.geometry[0][0])) {
+        // Note: the type is incorrect here
+        const geom = f.geometry as unknown as [number, number][][];
+        transformedGeom = geom.map((g1) =>
+          g1.map((g2) => {
+            const x = g2[0] * scale;
+            const y = g2[1] * scale;
+
+            if (bbox.minX > x) {
+              bbox.minX = x;
+            }
+
+            if (bbox.maxX < x) {
+              bbox.maxX = x;
+            }
+
+            if (bbox.minY > y) {
+              bbox.minY = y;
+            }
+
+            if (bbox.maxY < y) {
+              bbox.maxY = y;
+            }
+            return new Point(x, y);
+          })
+        );
+        numVertices = transformedGeom.reduce<number>(
+          (count, current) => count + current.length,
+          0
+        );
+      }
+      // Other feature types
+      else {
+        const geom = f.geometry as [number, number][];
+        transformedGeom = [
+          geom.map((g1) => {
+            const x = g1[0] * scale;
+            const y = g1[1] * scale;
+
+            if (bbox.minX > x) {
+              bbox.minX = x;
+            }
+
+            if (bbox.maxX < x) {
+              bbox.maxX = x;
+            }
+
+            if (bbox.minY > y) {
+              bbox.minY = y;
+            }
+
+            if (bbox.maxY < y) {
+              bbox.maxY = y;
+            }
+            return new Point(x, y);
+          })
+        ];
+        numVertices = transformedGeom.length;
+      }
+
+      if (f.type === 0) return null;
+
+      const geomType = {
+        [1]: GeomType.Point,
+        [2]: GeomType.Line,
+        [3]: GeomType.Polygon
+      }[f.type];
+
+      const feature: ProtomapsFeature = {
+        props: { ...(f.tags ?? {}) },
+        bbox,
+        geomType,
+        geom: transformedGeom,
+        numVertices
+      };
+
+      return feature;
+    })
+    .filter((f): f is ProtomapsFeature => f !== null);
```

Comment:
I see this has removed a cast, which is nice. Does the return type of the anonymous function passed to filter assert that the input is a ProtomapsFeature? Can't TypeScript figure that out without assertions if the the function passed to map on line 122 gets the return type annotation `ProtomapsFeature | null`? (Or the input type on this function gets a `ProtomapsFeature | null` type annotation.)